### ISSUE TITLE
[compiler] Compile error with Eclipse, but not with javac

### ISSUE THREAD

**Issue by jubax** [eclipse-jdt/eclipse.jdt.core#273]
The following code compiles with `javac`, but not with eclipse:
```
import java.util.ArrayList;
import java.util.List;

public class Instanceof<T> {

    interface A<T> {
	//
    }

    interface B<T> extends A<T> {
	//
    }

    interface Foo<T> {
	A<T> getX();
    }

    public void test() {
	List<? extends A<T>> list = new ArrayList<>();

	// Compile error in the following line: 
	//   "Type capture#1 of ? extends Instanceof.A<T> cannot be safely cast to Instanceof.A<T>
	if (list.get(0) instanceof B<T>) { // <== compile error with Eclipse, but not with javac
	    // do something
	}

	// Unchecked cast warnings
	if (((B<T>) list.get(0)) instanceof B<T>) {
	    // do something
	}

    }
}
```

Is there a workaround without changing the code?

**Comment by stephan-herrmann** [comment]
This code uses a feature introduced in our implementation via https://bugs.eclipse.org/bugs/show_bug.cgi?id=562392.

Perhaps @jarthana  can provide some insights?

**Comment by forax** [comment]
One of my student stumble on the same issue yesterday but with an instanceof and a type pattern.

The JLS section 15.20.2 [1] says
  The RelationalExpression must be downcast compatible with the ReferenceType (5.5 [1]), or a compile-time error occurs.

i.e. expr instanceof B<T> is allowed if (B<T>) expr is allowed.
 
The problem is that ecj considers that it's an unsafe cast
  List<? extends A<T>> list = new ArrayList<>();
  var v = (B<T>) list.get(0);   <-- unsafe cast
     
But it should not.

[1] https://docs.oracle.com/javase/specs/jls/se19/html/jls-5.html#jls-5.5

**Comment by jarthana** [comment]
Will consider this for 4.27.

**Comment by forax** [comment]
Thanks Jay !

**Comment by jarthana** [comment]
I am thinking of splitting these two conditions with the latter being used for issuing the unsafe cast warning:

`if (!isLegal || (this.bits & ASTNode.UnsafeCast) != 0) {`

@stephan-herrmann Stephan, I am a bit baffled by this testcase:

`    public void foo(A<? extends T> obj) {  
    	if (obj instanceof B<T> ) {
    		//   
    	} 
    } `

Javac issues an error here and with my above fix, we will only get the warning. I am using the existing `checkCastTypesCompatibility()` for this check. Is this expected that this method returns `true` for this code?

**Comment by jarthana** [comment]
Sorry, couldn't complete this in 4.27. Will keep it on my radar for 4.28.

**Comment by jarthana** [comment]
I spent a bite more time with this. My original work during the instanceof pattern matching basically rides on the existing framework for determining unsafe cast. One can see that `InstanceofExpression#checkUnsafeCast()` basically delegates it to `CastExpression#checkUnsafeCast()`. 

Let's alter the reported code a bit more to have a cast expression:
`B<T> bb = (B<T>) list.get(0);`

Now, ECJ produces an unchecked cast warning here. But no such warnings from Javac. So, my conclusion is, this is something that needs fixing in the original code in `checkUnsafeCast()` and `checkCastTypesCompatibility()`.

@stephan-herrmann , any tips from you for further investigation? I am out of depth here.

**Comment by srikanth-sankaran** [comment]
@jarthana if this can wait for a few weeks please assign to me - I'll work with @stephan-herrmann to wrap this up. Just that I have some travel coming up and may not be able to attend to it before next 2-3 weeks.

**Comment by jarthana** [comment]
> [@jarthana](https://github.com/jarthana) if this can wait for a few weeks please assign to me - I'll work with [@stephan-herrmann](https://github.com/stephan-herrmann) to wrap this up. Just that I have some travel coming up and may not be able to attend to it before next 2-3 weeks.

Yes, no problem. And thanks for the offer. I will assign this to you.
